### PR TITLE
Fix inconsistent negative subassembly indices between different sizeof(size_t)

### LINKED ISSUE #15953: Inconsistent subassembly object ID for different sizes of size_t
## Description

When compiling `solc` for different sizes of `size_t`, e.g., comparing emscripten and linux builds, the object ID of subassembly paths as determined here

https://github.com/ethereum/solidity/blob/a65229279d769a7b2936aef2aa802d77abbdc343/libevmasm/Assembly.cpp#L1825

will be inconsistent between the two builds. This potentially leads to inconsistent assembly text between emscripten and native 64 bit builds.

We should test the precise impact this has and fix it to the 64 bit behavior.

### PR DESCRIPTION
**Pull request by clonker**
When computing object ids for referencing subassemblies, these ids are currently determined as negative DFS enumeration based on `size_t`: 

https://github.com/ethereum/solidity/blob/a65229279d769a7b2936aef2aa802d77abbdc343/libevmasm/Assembly.cpp#L1825

This PR changes this to `uint64_t`, so that assembly text - in particular `PUSH #[$]` instructions - is consistent over different type sizes of `size_t`.

Since these indices are stored in a map and later referenced for exporting and importing, as well as (currently) `numeric_limits<size_t>::max()` is used to indicate the root object or an empty state which is used in, eg, various asserts, I found it helpful to wrap the `uint64_t` into a struct so that these places become apparent and dealing with the sizes is explicit. This caused a whole avalanche of changes, as such object ids (or more generally: `SubAssemblyID`s) are used in many places. In particular, also `yul::Object::subId`s are subject to the type.

We were lacking a cmdline test that has non-zero `PUSH #[$]`es, so I added one. Since emscripten builds do not execute them, we have no direct regression test by this but I have added one to the `solc-js` tests. While doing that I realized that it is currently not possible (to me at least :P) to request ASM json output via standard json interface if the language is Yul.

Fixes #15953

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `12b9a65822239f0daa1837f8e50b2670b3f8eb4d`:
```
Error: Trailing whitespace found:
Changelog.md:10:* Commandline Interface: Fix possible inconsistency in subassembly IDs between target architectures in `--asm-json` output. 

```
Please check that your changes are working as intended.

**Comment by r0qs** [comment]
>  While doing that I realized that it is currently not possible (to me at least :P) to request ASM json output via standard json interface if the language is Yul.
> 

Isn't that the `evm.assembly` option?

```json
{
	"language": "Yul",
	"sources":
	{
		"C":
		{
			"content": "{ let x := mload(0) sstore(add(x, 0), 0) }"
		}
	},
	"settings":
	{
		"outputSelection":
		{
			"*": { "*": ["evm.assembly"], "": [ "*" ] }
		}
	}
}
```

**Comment by clonker** [comment]
> Isn't that the `evm.assembly` option?

The equivalent that is showing the `PUSH #[$]`es would be `evm.legacyAssembly`

File: libevmasm/Assembly.h [hunk 1]
```diff
@@ -81,8 +82,16 @@ class Assembly
 	AssemblyItem newData(bytes const& _data) { util::h256 h(util::keccak256(util::asString(_data))); m_data[h] = _data; return AssemblyItem(PushData, h); }
 	bytes const& data(util::h256 const& _i) const { return m_data.at(_i); }
 	AssemblyItem newSub(AssemblyPointer const& _sub) { m_subs.push_back(_sub); return AssemblyItem(PushSub, m_subs.size() - 1); }
-	Assembly const& sub(size_t _sub) const { return *m_subs.at(_sub); }
-	Assembly& sub(size_t _sub) { return *m_subs.at(_sub); }
+	Assembly const& sub(SubAssemblyID const _sub) const
+	{
+		solAssert(_sub.value <= std::numeric_limits<size_t>::max());
+		return *m_subs.at(static_cast<size_t>(_sub.value));
+	}
```
Comment (cameel) [inline libevmasm/Assembly.h]: This is something that I'd implicitly assume to be true for such a type. I think that asserting it once in the definition of `SubAssemblyID` would be enough:

```c++
static_assert(std::numeric_limits<value_type>::max() <= std::numeric_limits<size_t>::max());
```

File: libevmasm/AssemblyItem.cpp [hunk 2]
```diff
@@ -352,13 +352,13 @@ std::string AssemblyItem::toAssemblyText(Assembly const& _assembly) const
 		break;
 	case PushTag:
 	{
-		size_t sub{0};
+		SubAssemblyID sub;
 		size_t tag{0};
 		std::tie(sub, tag) = splitForeignPushTag();
```
Comment (cameel) [inline libevmasm/AssemblyItem.cpp]: At first I wanted to point out that you're changing the initial value from `0` to `max()`, but then I noticed it's getting overwriten anyway. This would be clearer:

```suggestion
		auto [sub, tag] = splitForeignPushTag();
```

Comment (cameel) [reply libevmasm/AssemblyItem.cpp]: Same in the `tagRef` loop in `assembleLegacy()`.

File: libevmasm/Assembly.cpp [hunk 3]
```diff
@@ -1332,17 +1334,20 @@ LinkerObject const& Assembly::assembleLegacy() const
...
 				ret.bytecode.resize(ret.bytecode.size() + bytesPerDataRef);
 			});
 			break;
 		case PushSubSize:
 		{
 			assembleInstruction([&](){
-				assertThrow(item.data() <= std::numeric_limits<size_t>::max(), AssemblyException, "");
-				auto s = subAssemblyById(static_cast<size_t>(item.data()))->assemble().bytecode.size();
+				assertThrow(item.data() <= std::numeric_limits<SubAssemblyID::value_type>::max(), AssemblyException, "");
+				auto s = subAssemblyById({static_cast<SubAssemblyID::value_type>(item.data())})->assemble().bytecode.size();
```
Comment (cameel) [inline libevmasm/Assembly.cpp]: This assembling code would be more concise if we defined a conversion method in `SubAssemblyID()` that does the cast and asserts that the value is in the range of the `value_type`.

We could then use that in several other places, e.g. `CompilerContext.h` or `EthAssemblyAdapter.cpp`.

File: libevmasm/Assembly.cpp [hunk 4]
```diff
@@ -1796,47 +1801,46 @@ LinkerObject const& Assembly::assembleEOF() const
...
 }
 
-Assembly const* Assembly::subAssemblyById(size_t _subId) const
+Assembly const* Assembly::subAssemblyById(SubAssemblyID const _subId) const
 {
-	std::vector<size_t> subIds = decodeSubPath(_subId);
+	std::vector<SubAssemblyID> subIDs = decodeSubPath(_subId);
 	Assembly const* currentAssembly = this;
-	for (size_t currentSubId: subIds)
+	for (auto [subIDIndex]: subIDs)
```
Comment (cameel) [inline libevmasm/Assembly.cpp]: Wait, isn't this an ID rather than an index? You're iterating a `vector` of `SubAssemblyID`s. 
Though I also see that it gets cast to `size_t` without extracting the `.value`. Why does that even work?

This would be much clearer without `auto` obscuring the type.

```suggestion
	for (SubAssemblyID [subID]: subIDs)
```

File: libevmasm/AssemblyItem.cpp [hunk 5]
```diff
@@ -50,7 +50,7 @@ std::string toStringInHex(u256 _value)
 
 }
 
-AssemblyItem AssemblyItem::toSubAssemblyTag(size_t _subId) const
+AssemblyItem AssemblyItem::toSubAssemblyTag(SubAssemblyID _subId) const
 {
 	assertThrow(data() < (u256(1) << 64), util::Exception, "Tag already has subassembly set.");
```
Comment (cameel) [inline libevmasm/AssemblyItem.cpp:55]: Is this assert correct? I think it will fail when `_subID` is zero, but zero is a valid ID and an empty one is represented with `max()` instead. The message indicates that the function considers `max()` to be set and `0` to be unset, which does not seem right.

File: libsolidity/codegen/CompilerContext.h [hunk 6]
```diff
@@ -227,9 +229,9 @@ class CompilerContext
 	/// on the stack. @returns the pushsub assembly item.
 	evmasm::AssemblyItem addSubroutine(evmasm::AssemblyPointer const& _assembly) { return m_asm->appendSubroutine(_assembly); }
 	/// Pushes the size of the subroutine.
-	void pushSubroutineSize(size_t _subRoutine) { m_asm->pushSubroutineSize(_subRoutine); }
+	void pushSubroutineSize(evmasm::SubAssemblyID _subRoutine) { m_asm->pushSubroutineSize(_subRoutine); }
 	/// Pushes the offset of the subroutine.
-	void pushSubroutineOffset(size_t _subRoutine) { m_asm->pushSubroutineOffset(_subRoutine); }
+	void pushSubroutineOffset(evmasm::SubAssemblyID _subRoutine) { m_asm->pushSubroutineOffset(_subRoutine); }
```
Comment (cameel) [inline libsolidity/codegen/CompilerContext.h:232]: Oh, it's the first time I see assemblies referred to as subroutines. It really stretches the definition. From Wikipedia:

> In computer programming, a function (also procedure, method, subroutine, routine, or subprogram) is a callable unit[1] of software logic that has a well-defined interface and behavior and can be invoked multiple times. 

It's going to become very confusing with EOF, where we have both functions (code sections) and assemblies (containers). I'd by default interpret "subroutine" to mean to former (there was even a competing "simple subroutines" EIP). We should really rename this at some point.

File: libyul/backends/evm/EthAssemblyAdapter.h [hunk 7]
```diff
@@ -88,6 +88,6 @@ class EthAssemblyAdapter: public AbstractAssembly
 
 	evmasm::Assembly& m_assembly;
 	std::map<SubID, u256> m_dataHashBySubId;
-	size_t m_nextDataCounter = std::numeric_limits<size_t>::max() / 2;
+	SubID::value_type m_nextDataCounter = std::numeric_limits<SubID::value_type>::max() / 2;
```
Comment (cameel) [inline libyul/backends/evm/EthAssemblyAdapter.h]: Ugh, I thought I understood what's happening and then I ran into this. WTF are we doing? It just makes my head hurt. If I understand correctly:
- `Assembly` keeps track of chunks of data and subassemblies separately.
    - Assemblies are stored in `m_subs` vector.
    - Data is stored by hash in `m_data`.
        - Except for metadata, which is singled out and stored in `m_auxiliaryData`.
- `EthAssemblyAdapter` holds the assembly and independently keeps track of all the appended data chunks in `m_dataHashBySubId`.
    - It gives them fake IDs in the upper half of the subassembly ID space
- When a Yul object is being compiled, `EVMObjectCompiler` [adds subassemblies and data chunks via the adapter](https://github.com/ethereum/solidity/blob/v0.8.29/libyul/backends/evm/EVMObjectCompiler.cpp#L58-L74), receiving both real and fake IDs.
    - It stores the IDs in `BuiltinContext`, each one associated with a Yul node by name.
    - This is done only for children of the object. Anything nested deeper does not get an ID.
- `BuiltinContext` is passed to Yul code transform and later to builtins defined in `EVMDialect`.
    - Builtins like `datasize` then [use `BuiltinContext` to map Yul node names to sub IDs](https://github.com/ethereum/solidity/blob/v0.8.29/libyul/backends/evm/EVMDialect.cpp#L297-L300).
        - If the name is present in `BuiltinContext`, they take the ID stored there.
        - If not, they assume the name is a dotted path and convert it to a sequence of IDs.
            - BTW, this is a pretty questionable assumption: [#13794](https://github.com/ethereum/solidity/issues/13794), [#15540](https://github.com/ethereum/solidity/issues/15540).
        - In either case the sequence is passed into [`EthAssemblyAdapter::appendDataSize()`](https://github.com/ethereum/solidity/blob/v0.8.29/libyul/backends/evm/EthAssemblyAdapter.cpp#L195-L205).
            - It assumes that if the first ID in the sequence is present in `m_dataHashBySubId`, it must be a fake ID assigned to a data chunk.
            - Otherwise it converts the path to a sub ID using [`Assembly::encodeSubPath()`](https://github.com/ethereum/solidity/blob/v0.8.29/libevmasm/Assembly.cpp#L1814-L1831) and assumes it must be a subassembly.
                - There seems to be an assumption that Yul `Object` tree structure matches the `Assembly` tree, which I'm not sure is true. For example we sometimes we move things (e.g. long Yul strings) to data chunks and the transition happens in evmasm optimizer, which means that they do not have corresponding `data` nodes at Yul level.
            - *I actually don't understand how paths to nested data are handled. I don't think they're are added to `BuiltinContext`, but then how do they not end up treated as assemblies by `appendDataSize()`?*

Overall, looks like we have several kinds of sub IDs sharing the same ID space, without any checks for conflicts (just assuming the space is large enough that they won't happen):
- Normal IDs assigned to subassemblies, starting at `0`.
    - EOF `ContainerID`s are a subset of this and are limited to `uint8_t`.
- Empty ID at `max()`.
- Fake data IDs starting at `max() / 2`.
- "Negative" IDs representing deeply nested paths, starting at `max()` and going down.

I think we should assign a non-overlapping region to each of them and have asserts enforcing that.

The way we assign IDs should also be more prominently documented. For example `m_subPaths` does not even say that the "negative" IDs mean.

File: libevmasm/Assembly.h (hunk 1 above)
Comment (cameel) [reply libevmasm/Assembly.h]: Also, isn't `uint64_t` implicitly convertible to `size_t`? You have static casts to `size_t` all over the place and I think they're not necessary.

File: libevmasm/AssemblyItem.cpp (hunk 5 above)
Comment (cameel) [reply libevmasm/AssemblyItem.cpp:55]: Also, now that the sub ID is well-defined as 64-bits, I think we should have asserts that the upper bits of data are actually zeros.

File: libevmasm/Assembly.cpp (hunk 4 above)
Comment (cameel) [reply libevmasm/Assembly.cpp]: You also have some other loops using `size_t` `subIDIndex` where just using `SubAssemblyID` for the counter would make things simpler. E.g. in `optimiseInternal()` I think you could just do this:
```c++
for (SubAssemblyID subID = 0; subID.value < m_subs.size(); ++subID.value)
```

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (cameel) [reply libevmasm/Assembly.cpp]: BTW, almost all of these `assertThrow()`s should really be `solAssert()`s. The few that do represent proper validations should be replaced with `solRequire()`.

I'd recommend converting any of them that you touch in your PRs.

File: libevmasm/Assembly.h [hunk 8]
```diff
@@ -47,9 +48,9 @@ using AssemblyPointer = std::shared_ptr<Assembly>;
 
 class Assembly
 {
-	using TagRefs = std::map<size_t, std::pair<size_t, size_t>>;
+	using TagRefs = std::map<size_t, std::pair<SubAssemblyID, size_t>>;
```
Comment (cameel) [inline libevmasm/Assembly.h:51]: We use `size_t` for tag ID but that should really be `uint64_t` as well. Would be nice to change that too at some point.

The tag+sub pair would also be better off as a proper struct with methods for converting to/from `u256`. We wouldn't then have to hard-code all those masks and 64s all over the place.

File: libevmasm/Assembly.h (hunk 1 above)
Comment (clonker) [reply libevmasm/Assembly.h]: This will fail for emscripten, as there `size_t` is 32 bits. :)

File: libevmasm/AssemblyItem.cpp (hunk 5 above)
Comment (clonker) [reply libevmasm/AssemblyItem.cpp:55]: > Is this assert correct? I think it will fail when `_subID` is zero, but zero is a valid ID and an empty one is represented with `max()` instead. The message indicates that the function considers `max()` to be set and `0` to be unset, which does not seem right.

You are not the only whose head hurts by this stuff :) the condition `data() < (u256(1) << 64)` checks just if `data()` fits into 64 bits, as we have

```cpp
u256(std::numeric_limits<uint64_t>::max()) + 1 == (u256(1) << 64)
```

So it would pass if `data()` is zero and it indeed already asserts that the upper bits are zero. Or am I misunderstanding your concern?

File: libevmasm/AssemblyItem.cpp (hunk 2 above)
Comment (clonker) [reply libevmasm/AssemblyItem.cpp]: fixed

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (clonker) [reply libevmasm/Assembly.cpp]: > if we defined a conversion method

I have added a conversion constructor, that is indeed a good idea, makes the code much more readable and the assertions more concentrated in the place where they're needed. 

> almost all of these `assertThrow()`s should really be `solAssert()`s.

Right, `solAssert` and/or `solRequire` is also much easier to digest than `assertThrow`

File: libevmasm/Assembly.cpp (hunk 4 above)
Comment (clonker) [reply libevmasm/Assembly.cpp]: It works because you can unpack structs much like tuples, `auto [subIDIndex] = subIDs.front()` already refers to the contained `value` member. Although `subIDIndex` was not a good name, should have been `subIDValue`.

In any case, I have changed the loop now to just iterate over the `SubAssemblyID` instances instead of contained members and then extract the index with a new `asIndex` method that does static casting and size checking if needed. As an added benefit it limits the amount of static casts being littered about.

File: libevmasm/Assembly.h (hunk 8 above)
Comment (clonker) [reply libevmasm/Assembly.h:51]: Agree, it would also make it easier to reason about the whole program and data flow.

File: libyul/backends/evm/EthAssemblyAdapter.h (hunk 7 above)
Comment (clonker) [reply libyul/backends/evm/EthAssemblyAdapter.h]: Sounds like a good idea to define non-overlapping regions. That way each of the regions can also be independently and cleanly documented. Perhaps in a follow up? 
Regarding documenting subpaths and negative IDs: @r0qs had written something up about it and iirc wanted to add it in some form to the documentation.

Comment (r0qs) [reply libyul/backends/evm/EthAssemblyAdapter.h]: > Sounds like a good idea to define non-overlapping regions. That way each of the regions can also be independently and cleanly documented. Perhaps in a follow up? Regarding documenting subpaths and negative IDs: @r0qs had written something up about it and iirc wanted to add it in some form to the documentation.

Yes, it is here: https://github.com/ethereum/solidity/wiki/Assembly-Indices

Comment (clonker) [reply libyul/backends/evm/EthAssemblyAdapter.h]: I have added a comment to `m_subPaths` that the subassembly IDs referenced there occupy the negative index space. I think a more thorough documentation of things should occur once we have defined non-overlapping regions. Then it's also straightforward where to put the documentation.

File: libevmasm/SubAssemblyID.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	using value_type = uint64_t;
+	SubAssemblyID() = default;
+	SubAssemblyID(value_type const _value): value(_value) {}
+	explicit SubAssemblyID(u256 const& _data)
+	{
+		solAssert(_data <= std::numeric_limits<value_type>::max());
+		value = static_cast<value_type>(_data);
+	}
+
+	value_type value = std::numeric_limits<value_type>::max();
```
Comment (cameel) [inline libevmasm/SubAssemblyID.h]: I's easier to see the data fields (which is usually the first thing I'm looking for in a struct of class) when they are clearly separated from functions rather than mixed in between the functions. I think it's a good convention to put them always at the beginning or the end (we already stick to it in most places).

File: libevmasm/Assembly.cpp [hunk 10]
```diff
@@ -798,20 +799,20 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 
 	// Run optimisation for sub-assemblies.
 	// TODO: verify and double-check this for EOF.
-	for (size_t subId = 0; subId < m_subs.size(); ++subId)
+	for (SubAssemblyID subID {0}; subID.value < m_subs.size(); ++subID.value)
```
Comment (cameel) [inline libevmasm/Assembly.cpp]:
```suggestion
	for (SubAssemblyID subID{0}; subID.value < m_subs.size(); ++subID.value)
```

Also in `AssemblyItem::splitForeignPushTag()`.

File: libevmasm/Assembly.cpp [hunk 11]
```diff
@@ -1258,10 +1259,10 @@ LinkerObject const& Assembly::assembleLegacy() const
 		if (item.type() == PushTag)
 		{
 			auto [subId, tagId] = item.splitForeignPushTag();
-			if (subId == std::numeric_limits<size_t>::max())
+			if (subId.empty())
 				continue;
-			assertThrow(subId < m_subs.size(), AssemblyException, "Invalid sub id");
-			auto subTagPosition = m_subs[subId]->m_tagPositionsInBytecode.at(tagId);
+			solRequire(subId.value < m_subs.size(), AssemblyException, "Invalid sub id");
```
Comment (cameel) [inline libevmasm/Assembly.cpp]: This looks more like like an assertion to me. I don't think there is any user input that we'd expect to trigger it.

```suggestion
			solAssert(subId.value < m_subs.size(), "Invalid sub id");
```

Same below, in `tagRefs` loop.

Comment (cameel) [reply libevmasm/Assembly.cpp]: Also, why remove them for `PushSub` and `PushSubSize` rather than turn into asserts? They seem still relevant and `SubAssemblyID` constructor will not reject an empty value.

File: libevmasm/BlockDeduplicator.h [hunk 12]
```diff
@@ -59,7 +60,7 @@ class BlockDeduplicator
 	static bool applyTagReplacement(
 		AssemblyItems& _items,
 		std::map<u256, u256> const& _replacements,
-		size_t _subID = size_t(-1)
+		SubAssemblyID _subID = {}
```
Comment (cameel) [inline libevmasm/BlockDeduplicator.h:63]: The comment above still mentions `-1`. Also has a typo: `insied`.

File: libevmasm/JumpdestRemover.cpp [hunk 13]
```diff
@@ -45,7 +45,7 @@ bool JumpdestRemover::optimise(std::set<size_t> const& _tagsReferencedFromOutsid
 			if (_item.type() != Tag)
 				return false;
 			auto asmIdAndTag = _item.splitForeignPushTag();
-			assertThrow(asmIdAndTag.first == std::numeric_limits<size_t>::max(), OptimizerException, "Sub-assembly tag used as label.");
+			solRequire(asmIdAndTag.first.empty(), OptimizerException, "Sub-assembly tag used as label.");
```
Comment (cameel) [inline libevmasm/JumpdestRemover.cpp]: I think that this one as well is just a general assumption rather than something that we expect to fail at runtime.

Comment (cameel) [reply libevmasm/JumpdestRemover.cpp]: These tags were actually always confusing to me, because I could not see in what situation we'd ever jump into another subassembly, so I dug into this now to find out. Turns out we never do. This mechanism seems to exist only to support internal function pointers in legacy codegen. Since we support storage variables of such pointer types and storage can be accessed both at runtime and at creation time, we actually combine two pointers in the slot: to a copy of the function that exists in the creation assembly and to one in the runtime assembly. This not new to me, but I only now realized that to support this we have to be able to refer to tags from the runtime assembly inside the creation assembly. I think that's the only situation where they'll appear.

Here's what it looks like in practice:

```solidity
contract C {
    function() fptr = f;
    function f() internal {}
}
```
```
EVM assembly:
  mstore(0x40, 0x80)
  or(tag_0_2, shl(0x20, tag_1))
  0x00
  ...
  sstore
  ...
tag_1:
  jump  // out
  ...
stop

sub_0: assembly {
    ...
    tag_2:
      jump      // out
    ...
}
```

Note that there's no tag named `tag_0_2`. It refers to `tag_2` in `sub_0`. I don't think we ever insert such tags into the assembly item stream (and that's what the `assertThrow()` is guarding against). We only push them on the stack using `PushTag`, but this does not trigger the assert.

Such tags could, of course, appear in assembly import - we should have a validation against them there - but here it should be only an assert.

File: libevmasm/AssemblyItem.cpp [hunk 14]
```diff
@@ -61,20 +61,21 @@ AssemblyItem AssemblyItem::toSubAssemblyTag(size_t _subId) const
 	return r;
 }
 
-std::pair<size_t, size_t> AssemblyItem::splitForeignPushTag() const
+std::pair<SubAssemblyID, size_t> AssemblyItem::splitForeignPushTag() const
 {
 	solAssert(m_type == PushTag || m_type == Tag || m_type == RelativeJump || m_type == ConditionalRelativeJump);
```
Comment (cameel) [inline libevmasm/AssemblyItem.cpp:66]: Based on my investigation, I think we could limit it to just `PushTag`:
```suggestion
	solAssert(m_type == PushTag);
```

And similarly in `setPushTagSubIdAndTag()` and `toSubAssemblyTag()`.

Though maybe it's better done in a separate PR. There could be other places like this and it would be good to also have assembly import validations against it.

File: libyul/backends/evm/EVMDialect.cpp [hunk 15]
```diff
@@ -292,10 +292,10 @@ std::vector<std::optional<BuiltinFunctionForEVM>> createBuiltins(langutil::EVMVe
 				_assembly.appendAssemblySize();
 			else
 			{
-			std::vector<size_t> subIdPath =
+			std::vector<AbstractAssembly::SubID> subIdPath =
```
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp]: Why not just import `SubassemblyID.h` and use the type directly?

File: test/cmdlineTests/asm_json_nested_structure/args
Comment (cameel) [inline test/cmdlineTests/asm_json_nested_structure/args:1]: `asm_json_nested_structure` does not sound very informative regarding the point of the test. How about `asm_json_subassembly_id_representation`?

It would also be nice to have a comment in the test explaining that references to subs nested deeper than one level are encoded as negative numbers and we want to make sure that they consistently take up 64-bits across platforms. It's clear in the PR now, but later when looking at the test without that context it will be less so.

Same about naming and comments in the solc-js test.

Comment (cameel) [reply test/cmdlineTests/asm_json_nested_structure/args:1]: I'd also add a test covering `Tag`s with non-zero sub ID since that path was modified by the PR and I'm not sure we have much coverage there. I think that an evmasm test based on #16021 would be enough for that.

File: test/externalTests/solc-js/asm-json.js [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+            },
+            sources: {
+                C: {
+                    content: fs.readFileSync(testdir + 'code_access_runtime.sol', 'utf8')
+                }
+            }
+        })));
+        st.ok(output);
+
+        function containsTarget(obj, target) {
```
Comment (cameel) [inline test/externalTests/solc-js/asm-json.js]:
```suggestion
        function containsAssemblyItem(assemblyJSON, item) {
```

Also, I think it would be best to limit the search to the `legacyAssembly` subobject. It seems that now you're searching the whole object and there could be random stuff there. It would also make this check easier to understand - it took me a moment to understand what's happening here, because this is not at all what I expected.

Comment (cameel) [reply test/externalTests/solc-js/asm-json.js]: Also as a general remark for the future - we should at some point create a CI job that runs the emscripten binary with those tests that provide Standard JSON input/output. It would make this whole JS mess unnecessary...

File: Changelog.md [hunk 17]
```diff
@@ -17,6 +17,7 @@ Compiler Features:
 
 
 Bugfixes:
+* Commandline Interface: Fix possible inconsistency in subassembly IDs between target architectures in `--asm-json` output.
```
Comment (cameel) [inline Changelog.md]: This is not just the interface:

```suggestion
* Assembler: Fix different IDs being assigned to subassemblies nested more than one level away, resulting in inconsistent `--asm-json` output between target architectures.
```

Comment (cameel) [reply Changelog.md]: Also, this needs to be moved to 0.8.31 now.

File: test/cmdlineTests/asm_json_nested_structure/args
Comment (cameel) [reply test/cmdlineTests/asm_json_nested_structure/args:1]: Please also add `--json-indent 4` in `args` (we should just make 4 the default some day).

File: libevmasm/SubAssemblyID.h (hunk 9 above)
Comment (cameel) [reply libevmasm/SubAssemblyID.h]: Also:

```suggestion
	ValueType value = std::numeric_limits<ValueType>::max();
```

**Review by cameel — COMMENTED** [review]
Main things needed here are to replace validations with asserts and adjust tests a bit. Other than that mostly minor stuff.

This should be enough for the fix, but it needs to be followed by a PR that properly splits the index space into regions (https://github.com/ethereum/solidity/pull/15955#discussion_r2055878749) and documents it. And introduce a type for the tag ID (https://github.com/ethereum/solidity/pull/15955#discussion_r2008606997).

File: libyul/backends/evm/EthAssemblyAdapter.h (hunk 7 above)
Comment (cameel) [reply libyul/backends/evm/EthAssemblyAdapter.h]: @r0qs I see that the doc does not mention the fake data IDs. Not sure if they can ever show up in assembly output, but even if they can't, they still affect it, because we can never safely allocate other kinds of IDs in the region they take up. So they effectively limit the range of those other kinds IDs.

File: libevmasm/Assembly.h (hunk 1 above)
Comment (cameel) [reply libevmasm/Assembly.h]: Right. At least the extra constructors and functionsin `SubAssemblyID` removed most of this noise from the code. The diff actually looks pretty simple in most places in this file now.

File: libevmasm/AssemblyItem.cpp (hunk 5 above)
Comment (cameel) [reply libevmasm/AssemblyItem.cpp:55]: No, that's what what I meant. I was thrown off by the fact that the encoding of the ID is shifted by one. I guess it makes sense to have empty value at `0`, though -1 would work as well and I'm not sure it's worth the confusion it might create (until now I wasn't even aware of this difference).

File: libevmasm/Assembly.cpp (hunk 10 above)
Comment (clonker) [reply libevmasm/Assembly.cpp]: I really wish we could just settle on one style that has clang format support (or find a close enough approximation of the current style) and be done with it :) the amount of time and energy things like this eat up is unreasonable

File: libevmasm/Assembly.cpp (hunk 11 above)
Comment (clonker) [reply libevmasm/Assembly.cpp]: > Also, why remove them for `PushSub` and `PushSubSize` rather than turn into asserts? 

Honestly I don't recall. Perhaps I did have an assert in the constructor at some point or I figured that the assert can never be triggered. In any case, putting them back, it doesn't hurt to have them.

File: libevmasm/AssemblyItem.cpp (hunk 14 above)
Comment (clonker) [reply libevmasm/AssemblyItem.cpp:66]: Makes sense, thanks for investigating this!

File: libyul/backends/evm/EVMDialect.cpp (hunk 15 above)
Comment (clonker) [reply libyul/backends/evm/EVMDialect.cpp]: I did it this way because they are used as argument to an method in `AbstractAssembly` - which expects `AbstractAssembly::SubID`. We could of course get rid of that typedef entirely.

File: test/externalTests/solc-js/asm-json.js (hunk 16 above)
Comment (clonker) [reply test/externalTests/solc-js/asm-json.js]: > Also, I think it would be best to limit the search to the `legacyAssembly` subobject. It seems that now you're searching the whole object and there could be random stuff there. It would also make this check easier to understand - it took me a moment to understand what's happening here, because this is not at all what I expected.

Fair enough, what did you expect then? In my mind it was fairly straightforward to search the result json for some target object and it also seemed reasonably unlikely to me to find the target in our specific case anywhere else in the output. It is not big deal in any case, I am changing it to only consider the `legacyAssembly` subobject.

> Also as a general remark for the future - we should at some point create a CI job that runs the emscripten binary with those tests that provide Standard JSON input/output. It would make this whole JS mess unnecessary...

Yes please!!

File: Changelog.md (hunk 17 above)
Comment (clonker) [reply Changelog.md]: > This is not just the interface:

Well yeah, although assuming the ID space isn't exhausted, it does behave consistently inside of the assembler. I don't mind either way so let's go with your phrasing. Although I would minimally amend it:

`* Assembler: Fix not using a fixed-width type for IDs being assigned to subassemblies nested more than one level away, resulting in inconsistent `--asm-json` output between target architectures.`

File: libyul/backends/evm/EthAssemblyAdapter.h (hunk 7 above)
Comment (r0qs) [reply libyul/backends/evm/EthAssemblyAdapter.h]: Indeed, that was not the initial intention of such documentation. However, I can extend it to cover your suggestion of defining non-overlapping regions for IDs.

File: libevmasm/Assembly.cpp (hunk 11 above)
Comment (clonker) [reply libevmasm/Assembly.cpp]: Now I recall :)

```cpp
solAssert(item.data() <= std::numeric_limits<size_t>::max());
```

this will fail for emscripten when `size_t` is just 32 bits. We have to change the assert to

```cpp
solAssert(item.data() <= std::numeric_limits<SubAssemblyID::ValueType>::max());
```

for it to be semantically the same. But this is trivially true.

File: test/cmdlineTests/asm_json_nested_structure/args
Comment (clonker) [reply test/cmdlineTests/asm_json_nested_structure/args:1]: done

**Review by r0qs — APPROVED** [review]
